ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -627,28 +638,208 @@ prop_generator_NonFailingBeararInfoScript (NonFailingBearerInfoScript s) = not (
 
 prop_client_server :: [ByteString] -> BearerInfoScript -> Property
 prop_client_server payloads (BearerInfoScript script) =
-    let tr = runSimTrace $ clientServerSimulation script' payloads
-    in -- Debug.traceShow script $
-       case traceResult True tr of
-         Left e         -> counterexample
-                             (unlines
-                               [ "=== Say Events ==="
-                               , unlines (selectTraceEventsSay' tr)
-                               , "=== Error ==="
-                               , show e ++ "\n"
-                               , "=== Trace Events ==="
-                               , unlines (show `map` traceEvents tr)
-                               ])
-                             False
-         Right Nothing  -> property False
-         Right (Just b) -> property b
+  prop_verify_connect_to_simulation sim
   where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim = clientServerSimulation script' payloads
+
     script' = toBearerInfo <$> script
 
+data TestError = UnexpectedOutcome
+  deriving (Show, Eq)
+
+instance Exception TestError
+
+prop_verify_connect_to_simulation
+  :: (forall s . IOSim s (Either SomeException ()))
+  -> Property
+prop_verify_connect_to_simulation sim =
+  let tr = runSimTrace sim
+   in case traceResult True tr of
+     Left e -> counterexample
+                (unlines
+                  [ "=== Say Events ==="
+                  , unlines (selectTraceEventsSay' tr)
+                  , "=== Trace Events ==="
+                  , unlines (show `map` traceEvents tr)
+                  , "=== Error ==="
+                  , show e ++ "\n"
+                  ])
```

Comment:
I am not sure I understood what you meant.

Reply:
If we have `counterexample (f1 tr) (counterexample (f2 tr) ...)` we must ensure that nighter `f1` nor `f2` throws an exception during evaluation.  The `selectTraceEventsSay'` (and friends) are used for that purpose; but if we avoid multiple `counterexamples` in the first place we didn't had this problem in the first place.  Though sometimes it might be more convenient to have multiple `counterexamples`.